heal/harvard_downloads.py

```diff
@@ -0,0 +1,268 @@
+and logged in to QDR after the redirect.
+
+The WTS-SERVER is a Gen3 commons that has been configured to return
+tokens for the 'idp' specified in the external_file_metadata.
+"""
+
+import os
+
+from pathlib import Path
+import requests
+from typing import Dict, List, Tuple
+from urllib.parse import unquote
+import zipfile
+
+from cdislogging import get_logger
+from gen3.auth import Gen3Auth
+from gen3.tools.download.drs_download import DownloadStatus, wts_get_token
+
+logger = get_logger("__name__", log_level="debug")
+
+
+def get_harvard_dataverse_files(
+    wts_hostname: str, auth, file_metadata_list: List, download_path: str = "."
+) -> Dict:
+    """
+    Retrieves external data from the Syracuse QDR.
+
+    Args:
+        wts_hostname (str): hostname for commons with wts
+        auth (Gen3Auth): auth for commons with wts
+        file_metadata_list (List of Dict): list of studies or files
+        download_path (str): path to download files and unpack
+
+    Returns:
+        Dict of download status
+    """
+    if not Path(download_path).exists():
+        logger.critical(f"Download path does not exist: {download_path}")
+        return None
+
+    completed = {}
+    logger.debug(f"Input file metadata list={file_metadata_list}")
+
+    for file_metadata in file_metadata_list:
+        id = get_id(file_metadata)
+        if id is None:
+            logger.warning(
+                f"Could not find 'study_id' or 'file_id' in metadata {file_metadata}"
+            )
+            continue
+        logger.info(f"ID = {id}")
+        completed[id] = DownloadStatus(filename=id, status="pending")
+
+        download_url = get_download_url_for_harvard_dataverse(file_metadata)
+        if download_url is None:
+            logger.critical(f"Could not get download_url for {id}")
+            completed[id].status = "invalid url"
+            continue
+
+        logger.debug(f"Ready to send request to download_url: GET {download_url}")
+        downloaded_file = download_from_url(
+            harvard_url=download_url,
+            headers=None,
+            download_path=download_path,
+        )
+        if downloaded_file is None:
+            completed[id].status = "failed"
+            continue
+
+        if downloaded_file.endswith("zip"):
+            # unpack if download is zip file
+            try:
+                logger.debug(f"Ready to unpack {downloaded_file}.")
+                unpackage_object(filepath=downloaded_file)
+            except Exception as e:
+                logger.critical(f"{id} had an issue while being unpackaged: {e}")
+                completed[id].status = "failed"
+
+            completed[id].status = "downloaded"
+            # remove the zip file
+            Path(downloaded_file).unlink()
+        else:
+            completed[id].status = "downloaded"
+
+    if not completed:
+        return None
+    return completed
+
+
+def download_from_url(
```

Comment:
this function (any many other functions in this file) are identical to their siblings in the QDR retriever code, we should move them to a shared `utils.py` module and use from there to reduce code reduncency